Feature/62150 add leading character to pattern input to initiate search

Kharonus opened the pull request:
# Ticket
[OP#62150](https://community.openproject.org/wp/62150)
[OP#62258](https://community.openproject.org/wp/62258)

# What are you trying to accomplish?
- use `/` to indicate the search word and replacement range
- input changes within an existing token triggers token replacement

# What approach did you choose and why?
- amend pattern input component to behave accordingly to new specification

# Information
- this PR should give a first impression of the new interaction pattern, hence a pull preview is enabled
- please expect edge cases I might have missed
- Any implementation chunks inside the linked feature are supposedly NOT included yet.

Kharonus committed:
[#62510] change pattern input autocompletion

- https://community.openproject.org/wp/62150
- use `/` to indicate the search word and replacement range
- input changes within an existing token triggers token replacement

Kharonus committed:
[#62150] refinement of pattern input

- show "insert as text" only, if there is no longer any match
- update caption

brunopagno commented on frontend/src/stimulus/controllers/pattern-input.controller.ts:
```diff
@@ -166,13 +168,24 @@ export default class PatternInputController extends Controller {
+      return;
+    }
+
+    const parentNode = this.currentRange.startContainer.parentNode;
+    if (parentNode !== null && this.isToken(parentNode)) {
+      this.replaceToken(token, parentNode);
+    } else {
+      this.insertToken(token);
```
:green_circle: I like this behaviour better than the previous. Nice :clap:

brunopagno commented on frontend/src/stimulus/controllers/pattern-input.controller.ts:
```diff
@@ -189,10 +202,7 @@ export default class PatternInputController extends Controller {
-      if (range.startContainer.parentNode === this.contentTarget) {
-        this.currentRange = range;
-      }
+      this.currentRange = selection.getRangeAt(0);
```
The change here _could_ mean that we don't even need a `currentRange` local variable anymore. It was meant to keep track of a valid range (as in, a range that is not inside a token).

Worth checking if I'm not forgetting something else, though.

Kharonus replied:
Hmm, I thought we need this variable to maintain the range, when I get out of the input with `ArrowDown` to select a completion element? Is my assumption wrong?

brunopagno replied:
True that. I had forgotten about it :sweat_smile: 
Was just happy for a moment that we could remove some more core. but you're right.

brunopagno commented on frontend/src/stimulus/controllers/pattern-input.controller.ts:
```diff
@@ -263,44 +273,60 @@ export default class PatternInputController extends Controller {
-      wordRange.deleteContents();
-      wordRange.insertNode(token);
+    let pos = targetOffset - 1;
+    while (pos > -1 && textContent.charAt(pos) !== COMPLETION_CHARACTER) { pos -= 1; }
```
:green_circle: Also much much better than before

brunopagno reviewed (commented):
The PR is still in draft, is that intended?

But the code changes look correct to me. There's a potential change with the `this.currentRange` if you want to investigate further.

I also tested locally and it looks good

Kharonus committed:
[#62150] improve replacement behaviour

- if no match was found, replace /-indicated word or even existing token
  with the search word

psatyal reviewed (approved):
Did a visual check and I really like the new implementation. Great job, guys!

brunopagno reviewed (approved):
:ship:
